patches/tModLoader/Terraria/ModLoader/UI/Interface.cs

```diff
@@ -174,20 +174,27 @@ internal static void ModLoaderMenus(Main main, int selectedMenu, string[] button
 				ModLoader.DownloadedDependenciesOnStartup = true;
 
 				// Find dependencies that need to be downloaded.
-				var deps = ModOrganizer.IdentifyWorkshopDependencies().ToList();
-				bool promptDepDownloads = deps.Count != 0;
+				var missingDeps = ModOrganizer.IdentifyMissingWorkshopDependencies().ToList();
+				bool promptDepDownloads = missingDeps.Count != 0;
+
+				string message = $"{ModOrganizer.DetectModChangesForInfoMessage()}\n{string.Concat(missingDeps)}".Trim('\n');
 
-				string newDownloads = ModOrganizer.DetectModChangesForInfoMessage();
-				string dependencies = promptDepDownloads ? ModOrganizer.ListDependenciesToDownload(deps) : null;
-				string message = $"{newDownloads}\n{dependencies}".Trim('\n');
 				string cancelButton = promptDepDownloads ? Language.GetTextValue("tModLoader.ContinueAnyway") : null;
 				string continueButton = promptDepDownloads ? Language.GetTextValue("tModLoader.InstallDependencies") : "";
+
 				Action downloadAction = () => {
 					if (promptDepDownloads) {
-						//TODO: Would be nice if this used the names of the mods to replace the second x.ToString()
-						WorkshopHelper.SetupDownload(deps.Select(x => new ModDownloadItem(x.ToString(), x.ToString(), installed: null)).ToList(), previousMenuId: 0);
+						foreach (var slug in missingDeps) {
+							var item = WorkshopHelper.GetModDownloadItem(slug);
+							if (item == null) {
+								Utils.LogAndConsoleInfoMessage($"Could not find required mod dependency on Workshop: {slug}");
+								continue;
+							}
+
+							item.InnerDownloadWithDeps();
```

Comment:
This is a `Task` not a synchronous call..., all downloads need to be submitted simultaneously. Should be aggregated from `missingDeps`